# RadLex
92 colums

In [2]:
import pandas as pd

# Read the Excel file
df = pd.read_excel('/home/hbchen/Projects/MedKGC/resource/RadLex Term Browser.xls')

# Print the contents of the file
df

Class ID  \
0      http://www.radlex.org/RID/#RID43314   
1      http://www.radlex.org/RID/#RID10450   
2       http://www.radlex.org/RID/#RID7074   
3      http://www.radlex.org/RID/#RID22456   
4       http://www.radlex.org/RID/#RID7019   
...                                    ...   
46652  http://www.radlex.org/RID/#RID11406   
46653  http://www.radlex.org/RID/#RID30009   
46654   http://www.radlex.org/RID/#RID4747   
46655  http://www.radlex.org/RID/#RID23943   
46656  http://www.radlex.org/RID/#RID16558   

                                         Preferred Label  \
0                                   multiple lesions 2-5   
1                                       neutral position   
2                                       Brodmann area 13   
3      trunk of cutaneous branch of medial plantar nerve   
4                                       Brodmann area 22   
...                                                  ...   
46652                      gradient amplifier saturation   
46653                      upper anterior abdominal wall   
46654                                 Cowper syringocele   
46655  medial branch of posterior ramus of fourth cer...   
46656         neural network of mandibular part of mouth   

                                              Synonyms  \
0                                          2-5 lesions   
1                                              neutral   
2                             bA13|area 13 of Brodmann   
3                                                  NaN   
4      bA22|superior temporal area|area 22 of Brodmann   
...                                                ...   
46652                                              NaN   
46653                                              NaN   
46654                                              NaN   
46655                                              NaN   
46656                                              NaN   

                                             Definitions  Obsolete  CUI  \
0                                                    NaN       0.0  NaN   
1                                                    NaN       0.0  NaN   
2                                                    NaN       0.0  NaN   
3                                                    NaN       0.0  NaN   
4                                                    NaN       0.0  NaN   
...                                                  ...       ...  ...   
46652  Saturation of the phase encoding gradient ampl...       0.0  NaN   
46653                                                NaN       0.0  NaN   
46654                cystic dilation of the male urethra       0.0  NaN   
46655                                                NaN       0.0  NaN   
46656                                                NaN       0.0  NaN   

       Semantic Types                              Parents  AAL  ACR_ID  ...  \
0                 NaN  http://www.radlex.org/RID/#RID43313  NaN     NaN  ...   
1                 NaN  http://www.radlex.org/RID/#RID10420  NaN     NaN  ...   
2                 NaN  http://www.radlex.org/RID/#RID13125  NaN     NaN  ...   
3                 NaN  http://www.radlex.org/RID/#RID22442  NaN     NaN  ...   
4                 NaN  http://www.radlex.org/RID/#RID13125  NaN     NaN  ...   
...               ...                                  ...  ...     ...  ...   
46652             NaN  http://www.radlex.org/RID/#RID11403  NaN     NaN  ...   
46653             NaN  http://www.radlex.org/RID/#RID13405  NaN     NaN  ...   
46654             NaN   http://www.radlex.org/RID/#RID4743  NaN     NaN  ...   
46655             NaN  http://www.radlex.org/RID/#RID23942  NaN     NaN  ...   
46656             NaN  http://www.radlex.org/RID/#RID16554  NaN     NaN  ...   

       Surrounds                                          Synonym  \
0            NaN                                      2-5 lesions   
1            NaN                                          neutr

In [3]:
# print df 的列名和len，len of columns
print(df.columns)
print(len(df))
print(len(df.columns))

Index(['Class ID', 'Preferred Label', 'Synonyms', 'Definitions', 'Obsolete',
       'CUI', 'Semantic Types', 'Parents', 'AAL', 'ACR_ID',
       ...
       'Surrounds', 'Synonym', 'Synonym_German', 'Talairach', 'Term type',
       'Term_Status', 'Tributary_Of', 'UMLS_ID', 'UMLS_Term',
       'Unsanctioned Term'],
      dtype='object', length=103)
46657
103


## missing value analysis

In [4]:
# cal the percentage of missing values in each column
missing_values = df.isnull().mean() * 100

In [17]:
# 确保missing_values是一个Series
missing_values = pd.Series(missing_values)

# 按升序排序
missing_values_sorted = missing_values.sort_values()

# 设置pandas选项以显示所有行
pd.set_option('display.max_rows', None)

# 打印排序后的结果
print(missing_values_sorted)

# 如果你之后想要恢复默认设置，可以运行:
# pd.reset_option('display.max_rows')

Class ID                                            0.000000
Preferred Label                                     0.000000
Obsolete                                            0.002143
Parents                                             0.002143
http://data.bioontology.org/metadata/prefixIRI      0.004287
Is_A                                                0.053583
Preferred_name                                      2.003987
Preferred_name_German                               3.163512
FMAID                                              28.366590
Part_Of                                            55.211437
Regional_Part_Of                                   69.241486
Has_Subtype                                        70.645348
Synonyms                                           73.405920
Synonym                                            73.405920
Has_Part                                           86.820841
Continuous_With                                    88.794822
Constitutional_Part_Of  

## Source
254 sources

In [16]:
unique_sources = df['Source'].unique()

# print number of unique sources
print(unique_sources)

['LID' nan 'Playbook' 'LOINC' 'SIIM' 'PI-RADS' 'BI-RADS 5,'
 'Eisenberg, Ronald L. Atlas of Signs in Radiology.' 'LI-RADS' 'BI-RADS 5'
 'RadsWiki' 'Fleischner Society, BI-RADS 4' 'BI-RADS 4'
 'BI-RADS 4, BI-RADS 5' 'BI-RADS 5, BI-RADS 4, PI-RADS'
 'BI-RADS 5, BI-RADS 4, Playbook' 'Lung-RADS' 'breastNLP'
 'Radiology 1997; 204:765-768' 'LOINC, Playbook'
 'RadsWiki, Fleischner Society'
 'Radiology 2001; 218:230-232.  Fleischner Society' 'PLaybook'
 'Radiology 1999; 210:737-738' 'Radiology 2008; 247:291-292'
 'Radiology 2004; 230:109-110' 'ACR Actionable Reporting Work Group'
 "Wheeless' Textbook of Orthopaedics" 'LID, Playbook'
 'BI-RADS 5, BI-RADS 4' 'Radiology 1999; 212:541-542'
 'Radiology 2005; 236:653-654' 'MSKCC' 'user' 'PI-RADS2'
 'Radiology 2000; 215:263-265' 'Cancer Care Ontario'
 'Radiology 2007; 245:604-605' 'Radiology 2007; 244:927-928'
 'Fleischner Society' 'Radiology 2000; 216:428-429'
 'Radiology 2007; 243:297-298' 'Fleischner Society, Playbook'
 'Mulligan ME. Classic radio

In [17]:
# print("Number of unique sources:", len(unique_sources))

# Count the number of occurrences of each unique source
source_counts = df['Source'].value_counts()

# Print the whole counts
# print(source_counts.to_string())

import pandas as pd
pd.set_option('display.max_rows', None)
print(source_counts)

Source
Playbook                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             915
Eisenberg, Ronald L. Atlas of Signs in Radiology.                                                                                                                                                                                                                       

### Fleischner Society

# RadLex

## data prepossing

fenxi

In [ ]:
import pandas as pd
import os

def extract_columns(input_file, output_file):
    # Determine the file extension
    _, file_extension = os.path.splitext(input_file)
    
    # Read the Excel file
    if file_extension.lower() == '.xls':
        # For .xls files, we need to use the 'xlrd' engine
        df = pd.read_excel(input_file, engine='xlrd')
    else:
        # For .xlsx files, we can use the default engine
        df = pd.read_excel(input_file)

    # List of columns to extract
    columns_to_extract = [
        'Class ID',
        'Preferred Label',
        'Preferred_name',
        'Synonyms',
        'Synonym',
        'Definitions',
        'Definition',
        'Parents',
        'Is_A',
        'Continuous_With',
        'FMAID',
        'SNOMED_ID',
        'UMLS_ID',
        'Neurolex_ID',
        'http://data.bioontology.org/metadata/prefixIRI'
    ]

    # Extract only the specified columns that exist in the dataframe
    existing_columns = [col for col in columns_to_extract if col in df.columns]
    extracted_df = df[existing_columns]

    # Save the extracted data to a new Excel file
    extracted_df.to_excel(output_file, index=False, engine='openpyxl')

    print(f"Extracted data saved to {output_file}")

# Example usage
input_file = '/home/hbchen/Projects/MedKGC/resource/RadLex Term Browser.xls'
output_file = '/home/hbchen/Projects/MedKGC/resource/redlexX.xls'

extract_columns(input_file, output_file)

import pandas as pd

# Read the Excel file
df = pd.read_excel('/home/hbchen/Projects/MedKGC/resource/redlexX.xls')

# Print the contents of the file
df

Extracted data saved to /home/hbchen/Projects/MedKGC/resource/redlexX.xls


Class ID  \
0      http://www.radlex.org/RID/#RID43314   
1      http://www.radlex.org/RID/#RID10450   
2       http://www.radlex.org/RID/#RID7074   
3      http://www.radlex.org/RID/#RID22456   
4       http://www.radlex.org/RID/#RID7019   
...                                    ...   
46652  http://www.radlex.org/RID/#RID11406   
46653  http://www.radlex.org/RID/#RID30009   
46654   http://www.radlex.org/RID/#RID4747   
46655  http://www.radlex.org/RID/#RID23943   
46656  http://www.radlex.org/RID/#RID16558   

                                         Preferred Label  \
0                                   multiple lesions 2-5   
1                                       neutral position   
2                                       Brodmann area 13   
3      trunk of cutaneous branch of medial plantar nerve   
4                                       Brodmann area 22   
...                                                  ...   
46652                      gradient amplifier saturation   
46653                      upper anterior abdominal wall   
46654                                 Cowper syringocele   
46655  medial branch of posterior ramus of fourth cer...   
46656         neural network of mandibular part of mouth   

                                          Preferred_name  \
0                                   multiple lesions 2-5   
1                                       neutral position   
2                                       Brodmann area 13   
3      trunk of cutaneous branch of medial plantar nerve   
4                                       Brodmann area 22   
...                                                  ...   
46652                      gradient amplifier saturation   
46653                      upper anterior abdominal wall   
46654                                 Cowper syringocele   
46655  medial branch of posterior ramus of fourth cer...   
46656         neural network of mandibular part of mouth   

                                              Synonyms  \
0                                          2-5 lesions   
1                                              neutral   
2                             bA13|area 13 of Brodmann   
3                                                  NaN   
4      bA22|superior temporal area|area 22 of Brodmann   
...                                                ...   
46652                                              NaN   
46653                                              NaN   
46654                                              NaN   
46655                                              NaN   
46656                                              NaN   

                                               Synonym  \
0                                          2-5 lesions   
1                                              neutral   
2                             bA13|area 13 of Brodmann   
3                                                  NaN   
4      bA22|superior temporal area|area 22 of Brodmann   
...                                                ...   
46652                                              NaN   
46653                                              NaN   
46654                                              NaN   
46655                                              NaN   
46656                                              NaN   

                                             Definitions  \
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3                                                    NaN   
4                                                    NaN   
...                                                  ...   
46652  Saturation of the phase encoding gradient ampl...   
46653                                                NaN   
46654                cystic dilation of the male urethra   
46655                                        

In [ ]:
df["Preferred Label"]
is_same = df["Parents"].equals(df["Is_A"])
if not is_same:
    for index, row in df.iterrows():
        if row["Parents"] != row["Is_A"]:
            # print both 
            print(f"Parents: {row['Parents']}, Is_A: {row['Is_A']}")

Parents: http://www.radlex.org/RID/#RID28839, Is_A: http://www.radlex.org/RID/#RID15849
Parents: http://www.radlex.org/RID/#RID28956, Is_A: http://www.radlex.org/RID/#RID15849
Parents: http://www.radlex.org/RID/#RID12307, Is_A: http://www.radlex.org/RID/#RID15849
Parents: http://www.radlex.org/RID/#RID12626, Is_A: http://www.radlex.org/RID/#RID15849
Parents: http://www.radlex.org/RID/#RID13014, Is_A: http://www.radlex.org/RID/#RID15849
Parents: http://www.radlex.org/RID/#RID12562, Is_A: http://www.radlex.org/RID/#RID15849
Parents: http://www.radlex.org/RID/#RID13269, Is_A: http://www.radlex.org/RID/#RID15849
Parents: http://www.radlex.org/RID/#RID12971, Is_A: http://www.radlex.org/RID/#RID15849
Parents: http://www.radlex.org/RID/#RID28839, Is_A: http://www.radlex.org/RID/#RID15849
Parents: http://www.radlex.org/RID/#RID13221, Is_A: http://www.radlex.org/RID/#RID15849
Parents: nan, Is_A: nan
Parents: http://www.radlex.org/RID/#RID12971, Is_A: http://www.radlex.org/RID/#RID15849
Parents:

In [ ]:
is_same = df['Synonyms'].equals(df['Synonym'])
print(is_same)
if not is_same:
    for index, row in df.iterrows():
        if row['Synonyms'] != row['Synonym']:
            # print both 
            print(f"Synonyms: {row['Synonyms']}, Synonym: {row['Synonym']}")

True


In [ ]:
df["Preferred Label"]

0                                     multiple lesions 2-5
1                                         neutral position
2                                         Brodmann area 13
3        trunk of cutaneous branch of medial plantar nerve
4                                         Brodmann area 22
                               ...                        
46652                        gradient amplifier saturation
46653                        upper anterior abdominal wall
46654                                   Cowper syringocele
46655    medial branch of posterior ramus of fourth cer...
46656           neural network of mandibular part of mouth
Name: Preferred Label, Length: 46657, dtype: object

In [ ]:
df["Preferred Label"].isnull().any()

False

In [ ]:
df[df["Preferred Label"] == "lung"]


Class ID Preferred Label Preferred_name  \
22437  http://www.radlex.org/RID/#RID1301            lung           lung   

      Synonyms Synonym                                        Definitions  \
22437      NaN     NaN  Either of the pair of organs occupying the cav...   

                                              Definition  \
22437  Either of the pair of organs occupying the cav...   

                                   Parents  \
22437  http://www.radlex.org/RID/#RID13419   

                                      Is_A Continuous_With   FMAID SNOMED_ID  \
22437  http://www.radlex.org/RID/#RID13419             NaN  7195.0       NaN   

        UMLS_ID  Neurolex_ID http://data.bioontology.org/metadata/prefixIRI  
22437  C0024109          NaN                                        RID1301

## Embedding construction
from umls, radlex, snomed.

## hybrid search
### morphological search unsing levenshtein distance

In [ ]:
import pandas as pd
import numpy as np
from Levenshtein import ratio

def safe_levenshtein_ratio(query, x):
    """
    安全地计算Levenshtein比率，处理非字符串输入。
    """
    if pd.isna(x) or not isinstance(x, str):
        return 0.0  # 对于非字符串或NaN值，返回0相似度
    try:
        return ratio(str(query), str(x))
    except Exception:
        return 0.0  # 如果计算失败，返回0相似度

def find_similar_entities(query, df, column="Preferred Label", top_n=5):
    """
    查找与输入查询最相似的实体。
    
    参数:
    query (str): 要查找的实体名称
    df (pandas.DataFrame): 包含实体名称的DataFrame
    column (str): 包含实体名称的列名
    top_n (int): 返回的最相似实体数量
    
    返回:
    list of tuples: 包含(实体名称, 相似度分数)的列表,按相似度降序排序
    """
    # 确保查询是字符串
    query = str(query)
    
    # 计算查询与每个实体名称的Levenshtein比率
    similarities = df[column].apply(lambda x: safe_levenshtein_ratio(query, x))
    
    # 创建包含实体名称和相似度分数的DataFrame
    result_df = pd.DataFrame({
        'entity': df[column],
        'similarity': similarities
    })
    
    # 按相似度降序排序并返回top_n个结果
    return result_df.sort_values('similarity', ascending=False).head(top_n).values.tolist()

# 使用示例
# 假设df是您的DataFrame,包含"Preferred Label"列
query = "lung"

try:
    similar_entities = find_similar_entities(query, df)
    print(f"与'{query}'最相似的实体:")
    for entity, similarity in similar_entities:
        print(f"{entity}: {similarity:.2f}")
except Exception as e:
    print(f"发生错误: {e}")
    print("DataFrame的前几行:")
    print(df[["Preferred Label"]].head().to_string())
    print("\nPreferred Label列的数据类型:")
    print(df["Preferred Label"].dtype)
    print("\n非字符串值的示例 (如果有):")
    non_string = df[df["Preferred Label"].apply(lambda x: not isinstance(x, str))]
    print(non_string["Preferred Label"].head().to_string())

与'lung'最相似的实体:
lung: 1.00
lungs: 0.89
long: 0.75
lumen: 0.67
sling: 0.67


### semantic search

#### word_centric_embedding

In [ ]:
# 定义项目
items = [
    "Pneumothorax is An accumulation of air or gas in the PLEURAL CAVITY, which may occur spontaneously or as a result of trauma or a pathological process. The gas may also be introduced deliberately during PNEUMOTHORAX, ARTIFICIAL. (MSH)",
    "Lungs is Either of the pair of organs occupying the cavity of the thorax that effect the aeration of the blood. (MSH)",
    "Pleural is Of or pertaining to the pleura. (NCI)",
    "Thyroidectomy is Surgical removal of the thyroid gland."
]

mention = "Pneumothorax"

In [ ]:
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity
from scipy.spatial.distance import euclidean
from transformers import AutoTokenizer, AutoModel
import torch

# 首先，我们需要重新定义get_word_centric_embedding函数
def get_word_centric_embedding(text, target_word, model_name="bert-base-uncased"):
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    model = AutoModel.from_pretrained(model_name)

    inputs = tokenizer(text, return_tensors="pt", add_special_tokens=True)
    
    target_word_ids = tokenizer.encode(target_word, add_special_tokens=False)
    
    target_positions = []
    for i in range(len(inputs['input_ids'][0]) - len(target_word_ids) + 1):
        if inputs['input_ids'][0][i:i+len(target_word_ids)].tolist() == target_word_ids:
            target_positions.extend(range(i, i+len(target_word_ids)))

    with torch.no_grad():
        outputs = model(**inputs)

    word_embeddings = outputs.last_hidden_state[0, target_positions, :]
    word_centric_embedding = torch.mean(word_embeddings, dim=0)

    return word_centric_embedding.numpy()

# 定义相似性计算函数
def cosine_sim(a, b):
    return cosine_similarity([a], [b])[0][0]

def euclidean_sim(a, b):
    return 1 / (1 + euclidean(a, b))

def dot_product_sim(a, b):
    return np.dot(a, b)

# 获取每个项目的target word和embedding
item_embeddings = []
for item in items:
    target_word = item.split(' is')[0]
    embedding = get_word_centric_embedding(item, target_word)
    item_embeddings.append(embedding)

# 获取mention的embedding
mention_embedding = get_word_centric_embedding(mention, mention)


# 计算相似性并排序
similarity_methods = {
    "Cosine Similarity": cosine_sim,
    "Euclidean Similarity": euclidean_sim,
    "Dot Product": dot_product_sim
}

for method_name, sim_func in similarity_methods.items():
    print(f"\nUsing {method_name}:")
    similarities = [sim_func(mention_embedding, item_emb) for item_emb in item_embeddings]
    sorted_indices = np.argsort(similarities)[::-1]
    
    for i, idx in enumerate(sorted_indices):
        print(f"{i+1}. {items[idx][:50]}... (Similarity: {similarities[idx]:.4f})")

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]


Using Cosine Similarity:
1. Pneumothorax is An accumulation of air or gas in t... (Similarity: 0.6856)
2. Pleural is Of or pertaining to the pleura. (NCI)... (Similarity: 0.4017)
3. Thyroidectomy is Surgical removal of the thyroid g... (Similarity: 0.3494)
4. Lungs is Either of the pair of organs occupying th... (Similarity: 0.1902)

Using Euclidean Similarity:
1. Pneumothorax is An accumulation of air or gas in t... (Similarity: 0.1000)
2. Pleural is Of or pertaining to the pleura. (NCI)... (Similarity: 0.0741)
3. Thyroidectomy is Surgical removal of the thyroid g... (Similarity: 0.0691)
4. Lungs is Either of the pair of organs occupying th... (Similarity: 0.0546)

Using Dot Product:
1. Pneumothorax is An accumulation of air or gas in t... (Similarity: 86.9239)
2. Pleural is Of or pertaining to the pleura. (NCI)... (Similarity: 51.9061)
3. Thyroidectomy is Surgical removal of the thyroid g... (Similarity: 47.6269)
4. Lungs is Either of the pair of organs occupying th... (Similarity: 

这些结果很有趣，让我们来分析一下这三种相似度计算方法的表现：

1. 余弦相似度 (Cosine Similarity)：

余弦相似度测量两个向量之间的角度，范围通常在-1到1之间，1表示方向完全相同，0表示正交，-1表示方向相反。在我们的结果中，值域在0到1之间，这是因为我们使用的是非负的词嵌入。

优点：
- 对向量的大小不敏感，主要关注方向。
- 结果易于解释，0.6856的相似度表示相当高的相似性。
- 清晰地区分了最相关的项（Pneumothorax）和其他项。

观察：
- Pneumothorax和Pleural之间的相似度差距明显，这符合预期，因为Pleural（胸膜）与Pneumothorax（气胸）在医学上确实相关。
- Lungs（肺）的相似度最低，这可能是因为虽然肺与气胸有关，但在语义表示上差异较大。

2. 欧几里得相似度 (Euclidean Similarity)：

这里使用的是欧几里得距离的倒数，因此较小的距离会产生较大的相似度值。

优点：
- 考虑了向量在所有维度上的绝对差异。
- 对数值的绝对大小敏感。

观察：
- 总体趋势与余弦相似度一致，但数值范围较小（0.0546到0.1000）。
- 区分度似乎不如余弦相似度明显，特别是在较低相似度的项目之间。

3. 点积 (Dot Product)：

点积直接计算两个向量对应元素的乘积和，没有归一化。

优点：
- 计算简单快速。
- 保留了向量大小的信息。

观察：
- 结果的数值范围较大（32.2225到86.9239）。
- 排序结果与其他两种方法一致，表明它在这个场景下也能有效区分相似度。
- 但由于缺乏归一化，可能难以跨不同长度的文本或不同规模的嵌入进行比较。

总体分析：

1. 一致性：所有三种方法都正确地识别出Pneumothorax定义为最相似的项，这证实了我们的方法在捕捉语义相似性方面的有效性。

2. 排序一致性：三种方法给出的排序完全一致，这增强了我们对结果可靠性的信心。

3. 区分能力：
   - 余弦相似度在区分高相似度和低相似度项目方面表现最好。
   - 欧几里得相似度的数值范围较小，可能需要更精细的阈值来区分相似度级别。
   - 点积提供了最大的数值范围，但缺乏标准化可能使其难以在不同场景中设置统一的阈值。

4. 语义捕捉：结果显示Pleural（胸膜）比Lungs（肺）更相似于Pneumothorax（气胸），这在医学上是合理的，因为气胸直接涉及胸膜腔。

5. 应用建议：
   - 对于需要标准化比较的场景，余弦相似度可能是最佳选择。
   - 如果向量的大小信息很重要，点积可能更合适。
   - 欧几里得相似度在这个场景中表现不如其他两种方法明显，但在某些特定应用中可能有其优势。

6. 改进空间：考虑到Lungs与Pneumothorax的相关性在医学上应该较高，但在我们的结果中相似度最低，这可能表明我们的嵌入方法或相似度计算在捕捉某些语义关系方面还有改进空间。

这个分析为我们提供了有价值的见解，可以指导我们在实际应用中选择合适的相似度度量方法，并帮助我们理解如何进一步优化我们的语义表示和相似度计算方法。

In [ ]:
mention = "Collapsed lung"  

# 获取mention的embedding
mention_embedding = get_word_centric_embedding(mention, mention)

for method_name, sim_func in similarity_methods.items():
    print(f"\nUsing {method_name}:")
    similarities = [sim_func(mention_embedding, item_emb) for item_emb in item_embeddings]
    sorted_indices = np.argsort(similarities)[::-1]
    
    for i, idx in enumerate(sorted_indices):
        print(f"{i+1}. {items[idx][:50]}... (Similarity: {similarities[idx]:.4f})")

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertModel: ['cls.predictions.decoder.weight', 'cls.seq_relationship.weight', 'cls.predictions.transform.dense.weight', 'cls.predictions.bias', 'cls.predictions.transform.dense.bias', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.transform.LayerNorm.weight']
- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).



Using Cosine Similarity:
1. Lungs is Either of the pair of organs occupying th... (Similarity: 0.3196)
2. Pleural is Of or pertaining to the pleura. (NCI)... (Similarity: 0.2404)
3. Thyroidectomy is Surgical removal of the thyroid g... (Similarity: 0.0996)
4. Pneumothorax is An accumulation of air or gas in t... (Similarity: 0.0829)

Using Euclidean Similarity:
1. Pleural is Of or pertaining to the pleura. (NCI)... (Similarity: 0.0655)
2. Pneumothorax is An accumulation of air or gas in t... (Similarity: 0.0606)
3. Thyroidectomy is Surgical removal of the thyroid g... (Similarity: 0.0588)
4. Lungs is Either of the pair of organs occupying th... (Similarity: 0.0584)

Using Dot Product:
1. Lungs is Either of the pair of organs occupying th... (Similarity: 55.9129)
2. Pleural is Of or pertaining to the pleura. (NCI)... (Similarity: 32.0704)
3. Thyroidectomy is Surgical removal of the thyroid g... (Similarity: 14.0128)
4. Pneumothorax is An accumulation of air or gas in t... (Similarity: 

非常感谢您提供这个新的测试结果。让我们先分析结果，然后探讨可能的原因。

结果分析：

1. 余弦相似度：
   - Lungs（肺）排在第一位，相似度为0.3196。
   - Pleural（胸膜）排在第二位，相似度为0.2404。
   - Pneumothorax（气胸）排在最后，相似度仅为0.0829。

2. 欧几里得相似度：
   - Pleural排在第一位，相似度为0.0655。
   - Pneumothorax排在第二位，相似度为0.0606。
   - 四个项目的相似度差异较小。

3. 点积：
   - Lungs排在第一位，相似度为55.9129。
   - Pleural排在第二位，相似度为32.0704。
   - Pneumothorax排在最后，相似度为10.8493。

总体观察：
- 三种方法都没有将Pneumothorax（气胸）识别为与"Collapsed lung"最相似的项目，这是不符合预期的。
- Lungs和Pleural在不同方法中交替排在前两位。
- 相似度的绝对值普遍较低，特别是与之前"Pneumothorax"的结果相比。

原因分析：

1. 词汇差异：
   - "Collapsed lung"和"Pneumothorax"虽然在医学上指代相同的概念，但在词汇上完全不同。我们的模型可能没有捕捉到这两个术语之间的等价关系。

2. 上下文缺失：
   - 我们的方法仅使用了单个词或短语（"Collapsed lung"）来生成embedding，缺乏更广泛的上下文信息。这可能导致模型难以理解其完整的医学含义。

3. 预训练模型的限制：
   - 使用的预训练模型（如BERT）可能在医学术语方面的训练不足，特别是在处理同义词或相关概念时。

4. Embedding方法的局限性：
   - 我们的词中心化embedding方法可能不足以捕捉复杂的医学概念关系，特别是当涉及到同义词或相关术语时。

5. 数据集的局限性：
   - 我们的小型数据集（仅4个项目）可能不足以提供足够的上下文和比较基础。

6. 缺乏医学知识整合：
   - 我们的方法没有明确地整合专业的医学知识，这可能导致在处理专业术语时表现不佳。

改进建议：

1. 扩展数据集：增加更多相关的医学术语和定义，提供更丰富的上下文。

2. 使用医学专业预训练模型：考虑使用在医学文献上预训练的语言模型，如BioBERT或ClinicalBERT。

3. 整合同义词信息：在embedding过程中考虑医学术语的同义词关系。

4. 引入知识图谱：结合医学知识图谱来增强语义理解。

5. 改进embedding方法：考虑使用更复杂的embedding技术，如考虑更广泛上下文的句子或段落级embedding。

6. 微调模型：在特定的医学语料库上微调预训练模型。

7. 使用多步骤方法：首先识别关键医学概念，然后在这些概念的定义或描述上进行相似度计算。

这个测试结果揭示了在处理专业领域（如医学）的语义相似度时面临的挑战。它强调了在构建这类系统时需要考虑领域特定知识和更复杂的语言理解机制的重要性。通过实施上述改进建议，我们有望提高系统在处理医学术语和概念时的准确性和可靠性。

In [ ]:
mention = "An accumulation of air or gas in the PLEURAL CAVITY"  

# 获取mention的embedding
mention_embedding = get_word_centric_embedding(mention, mention)

for method_name, sim_func in similarity_methods.items():
    print(f"\nUsing {method_name}:")
    similarities = [sim_func(mention_embedding, item_emb) for item_emb in item_embeddings]
    sorted_indices = np.argsort(similarities)[::-1]
    
    for i, idx in enumerate(sorted_indices):
        print(f"{i+1}. {items[idx][:50]}... (Similarity: {similarities[idx]:.4f})")

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertModel: ['cls.predictions.decoder.weight', 'cls.seq_relationship.weight', 'cls.predictions.transform.dense.weight', 'cls.predictions.bias', 'cls.predictions.transform.dense.bias', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.transform.LayerNorm.weight']
- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).



Using Cosine Similarity:
1. Pneumothorax is An accumulation of air or gas in t... (Similarity: 0.6461)
2. Pleural is Of or pertaining to the pleura. (NCI)... (Similarity: 0.5663)
3. Thyroidectomy is Surgical removal of the thyroid g... (Similarity: 0.3663)
4. Lungs is Either of the pair of organs occupying th... (Similarity: 0.3429)

Using Euclidean Similarity:
1. Pneumothorax is An accumulation of air or gas in t... (Similarity: 0.0935)
2. Pleural is Of or pertaining to the pleura. (NCI)... (Similarity: 0.0844)
3. Thyroidectomy is Surgical removal of the thyroid g... (Similarity: 0.0689)
4. Lungs is Either of the pair of organs occupying th... (Similarity: 0.0591)

Using Dot Product:
1. Pneumothorax is An accumulation of air or gas in t... (Similarity: 85.3918)
2. Pleural is Of or pertaining to the pleura. (NCI)... (Similarity: 76.2790)
3. Lungs is Either of the pair of organs occupying th... (Similarity: 60.5782)
4. Thyroidectomy is Surgical removal of the thyroid g... (Similarity: 